lib/pbio/drv/usb/usb_ev3.c

```diff
@@ -52,500 +28,823 @@
+static int usb_addr_needs_setting;
+// The USB device address to use
+static uint8_t usb_addr;
+// The active USB configuration index (either 0 if unconfigured, or else 1)
+static uint8_t usb_config;
+// Data which remains to be sent for a CONTROL IN request
+static const uint32_t *setup_data_to_send;
+// Size of data remaining for a CONTROL IN
+static unsigned int setup_data_to_send_sz;
+// Used to send one (or a few) bytes back, for simple GET commands on EP0.
+static uint32_t setup_misc_tx_byte;
+// Whether the device is using USB high-speed mode or not
+static int is_usb_hs;
+
+// Buffers, used for different logical flows on the data endpoint
+static uint8_t ep1_rx_buf[PYBRICKS_EP_PKT_SZ_HS];
+static uint8_t ep1_tx_response_buf[PYBRICKS_EP_PKT_SZ_HS];
+static uint8_t ep1_tx_status_buf[PYBRICKS_EP_PKT_SZ_HS];
+static uint8_t ep1_tx_stdout_buf[PYBRICKS_EP_PKT_SZ_HS];
+
+// Buffer status flags
+static volatile int usb_rx_is_ready;
+static volatile int usb_tx_response_is_not_ready;
+static volatile int usb_tx_status_is_not_ready;
+static volatile int usb_tx_stdout_is_not_ready;
+
+// CPPI DMA support code
+
+// Host Packet Descriptor
+// The TI support library has hardcoded assumptions about the layout of these structures,
+// so we declare it ourselves here in order to control it as we wish.
+typedef struct {
+    hPDWord0 word0;
+    hPDWord1 word1;
+    hPDWord2 word2;
+    uint32_t buf_len;
+    void *buf_ptr;
+    void *next_desc_ptr;
+    uint32_t orig_buf_len;
+    void *orig_buf_ptr;
+} __attribute__((aligned(32))) usb_cppi_hpd;
+
+// We only use a hardcoded descriptor for each logical flow,
+// rather than dynamically allocating them as needed
+enum {
+    CPPI_DESC_RX,
+    CPPI_DESC_TX_RESPONSE,
+    CPPI_DESC_TX_STATUS,
+    CPPI_DESC_TX_STDOUT,
+    // the minimum number of descriptors we can allocate is 32,
+    // even though we do not use nearly all of them
+    CPPI_DESC_COUNT = 32,
 };
 
-// The descriptor string table.
-const unsigned char *const g_pStringDescriptors[] =
-{
-    g_pLangDescriptor,
-    g_pManufacturerString,
-    g_pProductString,
-    g_pSerialNumberString,
-    g_pControlInterfaceString,
-    g_pConfigString
+enum {
+    CPPI_RX_SUBMIT_QUEUE = 0,
 };
 
-extern const tUSBBuffer g_sTxBuffer;
-extern const tUSBBuffer g_sRxBuffer;
+// CPPI memory
+usb_cppi_hpd cppi_descriptors[CPPI_DESC_COUNT];
+uint32_t cppi_linking_ram[CPPI_DESC_COUNT];
+
+// Fill in the CPPI DMA descriptor to receive a packet
+static void usb_setup_rx_dma_desc() {
+    cppi_descriptors[CPPI_DESC_RX] = (usb_cppi_hpd) {
+        .word0 = {
+            .hostPktType = 0x10,
+        },
+        .word1 = {},
+        .word2 = {
+            .pktRetQueue = RX_COMPQ1,
+        },
+        .buf_len = PYBRICKS_EP_PKT_SZ_HS,
+        .buf_ptr = ep1_rx_buf,
+        .next_desc_ptr = 0,
+        .orig_buf_len = PYBRICKS_EP_PKT_SZ_HS,
+        .orig_buf_ptr = ep1_rx_buf,
+    };
+
+    __asm__ volatile ("" ::: "memory");
```

Comment:
```suggestion
    CP15ICacheFlushBuff((uint32_t)&cppi_descriptors[CPPI_DESC_RX], sizeof(cppi_descriptors[CPPI_DESC_RX]));
```

Is this what we are going for here?

Reply:
Doesn't that mean that we need this on every register read/write then?

In the Linux kernel, we have helper functions with various levels of memory barriers exactly for this reason.

Reply:
No, `HWREG` uses a `volatile` access. All `volatile` accesses are ordered with respect to each other. The only thing that is not ordered is `volatile` vs regular accesses.

Reply:
Ah, so the reasoning here is that we want all struct fields to actually be assigned (committed to memory) before assigning the _pointer_ to the struct to the register.